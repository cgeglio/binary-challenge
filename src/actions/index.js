export const login = user => ({
  type: 'LOGIN',
  user
})

export const logout = user => ({
  type: 'LOGOUT',
  user
})

export const addCards = cards => ({
  type: 'ADD_CARDS',
  cards
})

export const removeCards = cards => ({
  type: 'REMOVE_CARDS',
  cards
})

export const addQuestion = question => ({
  type: 'ADD_QUESTION',
  question
})

export const removeQuestion = question => ({
  type: 'REMOVE_QUESTION',
  question
})

export const addFortune = fortune => ({
  type: 'ADD_FORTUNE',
  fortune
})

export const removeFortune = fortune => ({
  type: 'REMOVE_FORTUNE',
  fortune
})

export const addReading = currentReading => ({
  type: 'ADD_READING',
  currentReading
})

export const removeReading = currentReading => ({
  type: 'REMOVE_READING',
  currentReading
})

export const addFavorite = favorite => ({
  type: 'ADD_FAVORITE',
  favorite
})

export const removeFavorite = favorite => ({
  type: 'REMOVE_FAVORITE',
  favorite
})

export const resetFavorites = favorites => ({
  type: 'RESET_FAVORITES',
  favorites
})
