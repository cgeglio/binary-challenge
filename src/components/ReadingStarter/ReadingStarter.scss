@import "../../mixins.scss";

.golden-text {
  height: 180px;
  position: absolute;
}

.home-page {
  @include flex();
  height: auto;
  position: relative;
}

.form-container {
  @include flex();
  justify-content: flex-start;
  margin-top: 100px;
}

.circle {
  background-color: white;
  border-radius: 50%;
  height: 400px;
  position: absolute;
  width: 400px;
}

.question-form {
  @include flex();
  justify-content: center;
  margin-top: 160px;
  position: absolute;
}

.question-form input {
  @include input();
  font-size: 1em;
  height: 50px;
  text-indent: .8em;
  width: 300px;
}

.ask-btn {
  @include buttons();
  height: 30px;
  width: 180px;
}

.question-form .error-message {
  margin-top: 155px;
}
