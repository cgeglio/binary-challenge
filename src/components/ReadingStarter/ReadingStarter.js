import React, { Component } from 'react';
import './ReadingStarter.scss';
import { connect } from 'react-redux';
import { addQuestion } from '../../actions'
import golden from '../../images/golden.png';

export class ReadingStarter extends Component {
  constructor() {
    super();
    this.state={question: '', error: null}
  }

  changeQuestion = event => {
    this.setState({question: event.target.value})
  }

  verifyInput = event => {
    event.preventDefault();
    let values = Object.values(this.state);
    values.includes('') ? this.setState({error: 'Please submit a question for your reading.'}) : this.startReading();
  }

  startReading = () => {
    this.props.updateQuestion(this.state.question);
    this.setState({question: '', error: null});
  }

  render() {
    return (
      <section className='home-page'>
        <img src={golden} alt="The Future Is Golden" className='golden-text'/>
      <section className='form-container'>
        <div className='circle'></div>
        <form className='question-form'>
          <input
            type='text'
            placeholder='Enter a question...'
            name='question'
            value={this.state.question}
            onChange={this.changeQuestion}
          />
          <button onClick={this.verifyInput} className='ask-btn'>ASK THE CARDS</button>
          <p className='error-message'>{this.state.error}</p>
        </form>
        </section>
      </section>
    )
  }
}

export const mapStateToProps = state => ({
  user: state.user
})

export const mapDispatchToProps = dispatch => ({
  updateQuestion: question => (dispatch(addQuestion(question)))
})

export default connect(mapStateToProps, mapDispatchToProps)(ReadingStarter)
