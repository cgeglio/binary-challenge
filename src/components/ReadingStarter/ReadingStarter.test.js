import React from 'react';
import { ReadingStarter, mapDispatchToProps, mapStateToProps } from './ReadingStarter';
import { addQuestion } from '../../actions';
import { shallow } from 'enzyme';

describe('ReadingStarter', () => {

  let wrapper;
  let mockUser;

  beforeEach(() => {
    mockUser = {name: 'Bithcuits', id: 9};
    wrapper = shallow(<ReadingStarter user={mockUser} />);
  });

  describe('ReadingStarter container/component', () => {
    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should start off with a default state', () => {
      expect(wrapper.state()).toEqual({question: '', error: null});
    });

    it('should set state with a question when changeQuestion is called', () => {
      const mockEvent = {
        target: {name: 'question', value: 'Is the sky blue?'},
        preventDefault: jest.fn()
      }
      wrapper.instance().changeQuestion(mockEvent);
      expect(wrapper.state()).toEqual({question: 'Is the sky blue?', error: null});
    });

    it('should verify question input when verifyInput is called and call startReading', () => {
      const mockEvent = { preventDefault: jest.fn() };
      wrapper.instance().startReading = jest.fn();
      wrapper.instance().setState({question: 'Is the sky blue?', error: null})

      wrapper.instance().verifyInput(mockEvent)
      expect(wrapper.instance().startReading).toHaveBeenCalled();
    });

    it('should set state with an error if there is not a question input', () => {
      const mockEvent = { preventDefault: jest.fn() };
      wrapper.instance().startReading = jest.fn();

      wrapper.instance().verifyInput(mockEvent)
      expect(wrapper.state('error')).toEqual('Please submit a question for your reading.');
    });

    it('should call updateQuestion and reset state when startReading is called', () => {
      const mockUpdate = jest.fn().mockImplementation();
      const expected = {question: '', error: null};
      wrapper = shallow(<ReadingStarter user={mockUser} updateQuestion={mockUpdate}/>);
      wrapper.instance().setState({question: 'Is the sky blue?'});

      wrapper.instance().startReading();
      expect(mockUpdate).toHaveBeenCalledWith('Is the sky blue?');
      expect(wrapper.state()).toEqual(expected);
    });

    it('should call verifyInput when ask the cards button is clicked', () => {
      wrapper.instance().verifyInput = jest.fn();
      wrapper.instance().forceUpdate()
      wrapper.find('.ask-btn').simulate('click')
      expect(wrapper.instance().verifyInput).toHaveBeenCalled();
    });
  });

  describe('mapStateToProps', () => {
    it('should return a user object', () => {
      const mockDispatch = jest.fn();
      wrapper = shallow(<ReadingStarter />);
      const mockState = {user: {name: 'Bithcuits', id: 9}, cards: [{name: 'The Magician', value: 8}]};
      const expected = {user: {name: 'Bithcuits', id: 9} };
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    })
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch with the addQuestion action when startReading is called', () => {
      const mockQuestion = 'Is the sky blue?';
      const mockDispatch = jest.fn();
      const actionToDispatch = addQuestion(mockQuestion);
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.updateQuestion(mockQuestion);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    })
  });
});
