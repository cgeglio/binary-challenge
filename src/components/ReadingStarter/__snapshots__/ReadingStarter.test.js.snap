// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReadingStarter ReadingStarter container/component should match the snapshot 1`] = `
<section
  className="home-page fade-in"
>
  <img
    alt="The Future Is Golden"
    className="golden-text"
    src="golden.png"
  />
  <section
    className="form-container"
  >
    <div
      className="circle"
    />
    <form
      className="question-form"
    >
      <select
        name="spread"
        onChange={[Function]}
      >
        <option>
          Select reading...
        </option>
        <option
          value={4}
        >
          4 Card Reading
        </option>
        <option
          value={3}
        >
          3 Card Reading
        </option>
        <option
          value={1}
        >
          1 Card Reading
        </option>
      </select>
      <input
        maxLength={55}
        name="question"
        onChange={[Function]}
        placeholder="Enter a question..."
        type="text"
        value=""
      />
      <button
        className="ask-btn"
        onClick={[Function]}
      >
        ASK THE CARDS
      </button>
      <p
        className="error-message"
      />
    </form>
  </section>
</section>
`;
